### PR TITLE
Block connections on config change

### PR DESCRIPTION
**Pull request by vlabo**

### CODE REVIEW

**Review by dhaavi — CHANGES_REQUESTED** [review]

**Review by dhaavi — CHANGES_REQUESTED** [review]
Almost there!
Please also fix the linter warning.

File: firewall/interception.go [hunk 1]
```diff
@@ -102,34 +102,29 @@ func interceptionPrep() error {
...
-		log.Errorf("failed to reset all connections: %q", err)
-	}
 
 	// reset all connection firewall handlers. This will tell the master to rerun the firewall checks
-	for _, id := range network.GetAllIDs() {
-		conn, err := getConnectionByID(id)
-		if err != nil {
-			continue
-		}
+	for _, conn := range network.GetAllConnections() {
```
[the remark this answers is not in the document]
Comment (vlabo) [reply firewall/interception.go:122]: are does just dns request for localhost or this includes the resolvers dns connections DoT, Doh ... ?
If it just localhost I agree there is no point of adding them

File: firewall/interception/nfq/conntrack.go [hunk 2]
```diff
@@ -17,37 +19,42 @@ func DeleteAllMarkedConnection() error {
...
+		for _, connection := range currentConnections {
+			err = nfct.Delete(ct.Conntrack, ct.IPv4, connection)
+			if err != nil {
+				numberOfErrors++
+			}
+		}
+
+		if numberOfErrors > 0 {
+			log.Warningf("nfq: failed to delete %d conntrack entries last error is: %s", numberOfErrors, err)
+		}
```
[the remark this answers is not in the document]
Comment (vlabo) [reply firewall/interception/nfq/conntrack.go:58]: Yes it was intentional, but now that I think about it it does not give a lot of value. I will remove it

File: network/connection_store.go [hunk 3]
```diff
@@ -52,9 +52,11 @@ func (cs *connectionStore) list() []*Connection {
...
 	defer cs.rw.RUnlock()
 
-	l := []*Connection{}
+	l := make([]*Connection, len(cs.items))
+	index := 0
 	for _, conn := range cs.items {
-		l = append(l, conn)
+		l[index] = conn
+		index++
 	}
```
Comment (dhaavi) [inline network/connection_store.go:60]: I find it easier to create an empty slice and the correct capacity and then just us append on it.
This way to code is easy to read and very safe.

Eg:
```
l := make([]*Connection, 0, len(cs.items))
for _, conn := range cs.items {
  l = append(l, conn)
}

File: firewall/interception/nfq/conntrack.go [hunk 4]
```diff
@@ -17,37 +19,42 @@ func DeleteAllMarkedConnection() error {
...
+		deleteMarkedConnections(nfct, ct.IPv6)
 	}
 
 	return nil
 }
 
-func getAllMarkedConnections(nfct *ct.Nfct, f ct.Family) []ct.Con {
+func deleteMarkedConnections(nfct *ct.Nfct, f ct.Family) {
 	// initialize variables
 	permanentFlags := [...]uint32{MarkAccept, MarkBlock, MarkDrop, MarkAcceptAlways, MarkBlockAlways, MarkDropAlways, MarkRerouteNS, MarkRerouteSPN}
```
[the remark this answers is not in the document]
Comment (dhaavi) [reply firewall/interception/nfq/conntrack.go:35]: Do you want to leave it as is? You did not comment or change something

**Review by dhaavi — CHANGES_REQUESTED** [review]

File: firewall/interception.go [hunk 5]
```diff
@@ -442,14 +485,31 @@ func initialHandler(conn *network.Connection, pkt packet.Packet) {
...
 	default:
+		conn.StopFirewallHandler()
+		issueVerdict(conn, pkt, 0, true)
+	}
+}
+
+func filterConnection(ctx context.Context, conn *network.Connection, pkt packet.Packet) {
+	if filterEnabled() {
+		log.Tracer(ctx).Trace("filter: starting decision process")
+		DecideOnConnection(ctx, conn, pkt)
```
Comment (vlabo) [inline firewall/interception.go:511]: Doesn't this crash when packet is `nil`?

File: firewall/interception.go [hunk 6]
```diff
@@ -559,17 +620,52 @@ func issueVerdict(conn *network.Connection, pkt packet.Packet, verdict network.V
...
+	switch {
+	case conn.Verdict.Active == network.VerdictUndecided:
+		// Apply first verdict without change.
+		conn.Verdict.Active = conn.Verdict.Firewall
+
+	case reference.IsPacketProtocol(conn.Entity.Protocol):
+		// For known packet protocols, apply firewall verdict unchanged.
+		conn.Verdict.Active = conn.Verdict.Firewall
+
+	case conn.Verdict.Active != conn.Verdict.Firewall:
```
Comment (vlabo) [inline firewall/interception.go:665]: Shouldn't there be a way of going from blocked to allowed?

File: firewall/interception.go (hunk 5 above)
Comment (dhaavi) [reply firewall/interception.go:511]: I think all checks correctly check for existence of pkt, as DNS requests don't have one either.

Comment (dhaavi) [reply firewall/interception.go:511]: I checked and made some improvements. `pkt` is hardly used.

**Review by dhaavi — APPROVED** [review]

**Pull request merged**